src/mesh/unstructured_mesh.C

```diff
@@ -1244,17 +1246,19 @@ void UnstructuredMesh::all_second_order (const bool full_ordered)
   if (!this->n_elem())
     return;
 
-  /*
-   * If the mesh is already second order
-   * then we have nothing to do.
-   * We have to test for this in a round-about way to avoid
-   * a bug on distributed parallel meshes with more processors
-   * than elements.
-   */
-  bool already_second_order = false;
-  if (this->elements_begin() != this->elements_end() &&
-      (*(this->elements_begin()))->default_order() != FIRST)
-    already_second_order = true;
+  // If every element in the range is already second order then we
+  // have nothing to do. Note: std::all_of() returns true for an empty
+  // range, which can happen for example in the DistributedMesh case
+  // when there are more processors than elements. In the case of an
+  // empty range we therefore we set already_second_order to false so
+  // that processors with empty ranges just "do whatever the other
+  // processors are doing".
+  bool already_second_order =
+    (range.begin() != range.end()) &&
+    std::all_of(range.begin(), range.end(),
+                [](const auto & elem){ return elem->default_order() == SECOND; });
+
+  // Check with other processors and possibly return early
   this->comm().max(already_second_order);
```

Comment:
We want `min()` now instead of `max()`, right?  If one processor has a range with all first-order elements that doesn't mean every processor does.

Also, if we use `min()` we can probably get rid of the separate empty-range check above - `std::all_of` should return `true` for an empty range.

Reply:
If any processor has a range with (some) first-order elements, then they can't all early return. This is a parallel-only function, or at least that was how I understood the original logic.

Reply:
The original logic wasn't logical.  I was assuming that all elements in the mesh were of the same order.

Reply:
This issue should be fixed now and I've also rebased on latest devel.